File: MPP/MPP_Iteratia3_C#/Common/ClientWorker.cs
using Common.model;
using Common.requests;
using Common.responses;
using Common.services;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Net.Sockets;
using System.Runtime.Serialization;
using System.Runtime.Serialization.Formatters.Binary;
using System.Text;
using System.Threading;
using System.Threading.Tasks;

namespace Common
{
    public class ClientWorker : IFestivalObserver
    {
        private IFestivalService server;
        private TcpClient client;

        private NetworkStream stream;
        private IFormatter formatter;
        private volatile bool connected;

        public ClientWorker(IFestivalService server, TcpClient client)
        {
            this.server = server;
            this.client = client;
            try
            {
                stream = client.GetStream();
                formatter = new BinaryFormatter();
                connected = true;
            }
            catch(Exception ex)
            {
                Console.WriteLine(ex.StackTrace);
            }
        }

        public void update(List<ConcertDTO> L)
        {
            try
            {
                sendResponse(new TicketSold(L));
            }
            catch(Exception e)
            {
                Console.WriteLine(e);
            }
        }

        public void run()
        {
            while (connected)
            {
                try
                {
                    object request = formatter.Deserialize(stream);
                    object response = handleRequest((Request)request);
                    if (response != null)
                    {
                        sendResponse((Response)response);
                    }
                }
                catch(Exception e)
                {
                    Console.WriteLine(e.StackTrace);
                }

                try
                {
                    Thread.Sleep(1000);
                }
                catch(Exception e)
                {
                    Console.Write(e.StackTrace + "\n");
                }
            }

            try
            {
                stream.Close();
                client.Close();
            }
            catch(Exception e)
            {
                Console.WriteLine("Error : " + e);
            }
        }

        private void sendResponse(Response response)
        {
            Console.WriteLine("Sending response : " + response);
            try
            {
                lock (stream)
                {
                    formatter.Serialize(stream, response);
                    stream.Flush();
                }
            }
            catch(Exception e)
            {
                Console.WriteLine(e.StackTrace);
            }
            
        }

        private object handleRequest(Request request)
        {
            Response resp = null;
            if(request is LoginRequest)
            {
                Console.WriteLine("Processing a login request ... ");
                LoginRequest logR = (LoginRequest)request;
                User u = logR.user;
                try
                {
                    Console.WriteLine("-- Sending request...");
                    lock (server)
                    { 
                        server.Login(u, this);
                    }
                    return new OkResponse("Utilizator logat ");
                }
                catch(Exception e)
                {
                    connected = false;
                    return new ErrorResponse("Credentiale invalide");
                }
            }

            if(request is LogoutRequest)
            {
                Console.WriteLine("Processing a logout request ... ");
                LogoutRequest logR = (LogoutRequest)request;
                User u = logR.user;
                try
                {
                    lock (server)
                    {
                        server.Logout(u, this);
                    }
                    connected = false;
                    return new OkResponse("User logged out");
                }
                catch(Exception e)
                {
                    return new ErrorResponse("Ceva nu-i ok");
                }
            }

            if(request is GetAllConcertsRequest)
            {
                Console.WriteLine("Processing request : get all concerts");
                try
                {
                    List<ConcertDTO> c = new List<ConcertDTO>();
                    lock (server)
                    {
                        c = server.getAllConcerts();
                    }
                    GetAllConcertsResponse r = new GetAllConcertsResponse(c);
                    sendResponse(r);
                    //return new GetAllConcertsResponse(c);
                }
                catch(Exception e)
                {
                    return new ErrorResponse("Ceva nu-i ok");
                }
            }

            if(request is FilterConcertsRequest)
            {
                Console.WriteLine("Processing request : filter concerts");
                try
                {
                    FilterConcertsRequest re = (FilterConcertsRequest)request;
                    List<ConcertDTO> c = new List<ConcertDTO>();
                    lock (server)
                    {
                        c = server.filterConcerts(re.data);
                    }
                    FilterConcertsResponse r = new FilterConcertsResponse(c);
                    sendResponse(r);
                    //return new FilterConcertsResponse(c);
                }
                catch(Exception e)
                {
                    return new ErrorResponse("nu-i ok");
                }
            }

            if(request is SellTicketRequest)
            {
                try
                {
                    //List<ConcertDTO> L = new List<ConcertDTO>();
                    SellTicketRequest r = (SellTicketRequest)request;
                    Bilet b = r.bilet;
                    lock (server)
                    {
                        server.sellTicket(b);
                    }
                    return new OkResponse("Vandut bilet");
                }
                catch(Exception e)
                {
                    Console.WriteLine(e);
                    return new ErrorResponse("ticket sold error");
                }
            }
            return resp;
        }
    }
}

File: MPP/MPP_Iteratia3_Java/Networking/src/main/java/ClientWorker.java

import Model.Bilet;
import Model.ConcertDTO;
import Model.User;
import requests.*;
import responses.*;

import java.io.IOException;
import java.io.ObjectInputStream;
import java.io.ObjectOutputStream;
import java.net.Socket;
import java.time.LocalDate;
import java.util.ArrayList;
import java.util.List;

public class ClientWorker implements Runnable, IObserver{
    private IService service;
    private Socket socket;
    private ObjectInputStream input;
    private ObjectOutputStream output;
    private volatile boolean connected;

    public ClientWorker(IService service, Socket socket){
        this.service = service;
        this.socket = socket;

        try{
            output = new ObjectOutputStream(socket.getOutputStream());
            output.flush();
            input = new ObjectInputStream(socket.getInputStream());
            connected = true;
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    public void run() {
        while(connected){
            try{
                Object req = input.readObject();
                Object resp = handleRequest((Request)req);
                if (resp!=null){
                    sendResponse((Response)resp);
                }
            }catch (Exception e){
                e.printStackTrace();
            }

            try {
                Thread.sleep(1000);
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
        }
        try {
            input.close();
            output.close();
            socket.close();
        } catch (IOException e) {
            System.out.println("Error "+e);
        }
    }

    private Object handleRequest(Request req) throws Exception {
        Response response = null;
        if(req instanceof LoginRequest){
            System.out.println(this.getClass()+ " Processing login request");
            LoginRequest lo = (LoginRequest) req;
            User u = lo.user;
            try{
                service.login(u, this);
                return new OkResponse(u.toString());
            }catch (Exception e){
                connected = false;
                return new ErrorResponse(e.getMessage());
            }
        }

        if (req instanceof LogoutRequest){
            LogoutRequest lo = (LogoutRequest)req;
            User u = lo.u;
            try{
                service.logout(u, this);
                try{
                    sendResponse(new OkResponse("Logout cu succes"));
                }catch (Exception e){
                    e.printStackTrace();
                }
                connected = false;
            }catch (Exception e){
                return new ErrorResponse(e.getMessage());
            }
        }

        if (req instanceof SellTicketRequest){
            System.out.println(this.getClass()+  " Processing sell ticket ");
            SellTicketRequest s = (SellTicketRequest)req;
            Bilet b = s.getBilet();
            try{
                service.sellTicket(b);
                return new OkResponse("TICKET sold");
            }catch (Exception e){
                return new ErrorResponse("Nu-i ok");
            }
        }

        if(req instanceof SearchConcertsRequest){
            SearchConcertsRequest sr = (SearchConcertsRequest)req;
            LocalDate d = sr.getData();
            List<ConcertDTO> rez = new ArrayList<>();
            try{
               rez= service.searchConcerts(d);
            }catch (Exception e){
                return new ErrorResponse(e.getMessage());
            }
            try{
                sendResponse(new SearchConcertsResponse(rez));
            }catch (Exception e){
                e.printStackTrace();
            }
        }

        if(req instanceof GetAllConcertsRequest){
            System.out.println(this.getClass() + " getting all the concerts ... method handleReq");
            GetAllConcertsRequest gc = (GetAllConcertsRequest)req;
            List<ConcertDTO> rez = null;
            try{
                rez = service.getAllConcerts();
            }catch (Exception e){
                e.printStackTrace();
            }
            try{
                sendResponse(new GetAllConcertsResponse(rez));
            }catch (Exception e){
                e.printStackTrace();
            }
            //return new GetAllConcertsResponse(rez);
        }

        /*if(req instanceof UpdateRequest){
            System.out.println("requests.Request to update all the clients");
            UpdateRequest u = (UpdateRequest) req;
            service.broadcastTable(u.l);
            return new TicketSoldResponse(u.l);
        }*/
        return response;
    }

    @Override
    public void update(List<ConcertDTO> l) {
        try{
            sendResponse(new TicketSoldResponse(l));
        }catch (Exception e){
            e.printStackTrace();
        }
    }

    private void sendResponse(Response resp) throws IOException {
        System.out.println("Throwing response : " + resp.getClass() + " from" + service.getClass());
        output.writeObject(resp);
        output.flush();
    }

}
